### PR TITLE
Glonass ephemeris

### PR DESCRIPTION
**Pull request by peddie**
This patch mainly adds support for RTCM v3 message number 1020 for GLONASS ephemerides, along with a few unrelated cleanups.

### CODE REVIEW

File: haskell/src/Data/RTCM3/Ephemeris.hs [hunk 1]
```diff
@@ -202,3 +207,244 @@ instance Binary Msg1019 where
...
+    _glonassEphemerisHeader_sat <- B.getWord8     6
+    _glonassEphemerisHeader_channel <- B.getWord8 5
+    return GlonassEphemerisHeader {..}
+
+  putBits _n GlonassEphemerisHeader {..} = do
+    B.putWord16be 12 _glonassEphemerisHeader_num
+    B.putWord8    6  _glonassEphemerisHeader_sat
+    B.putWord8    5  _glonassEphemerisHeader_channel
+
+data GlonassEphemeris = GlonassEphemeris
```
Comment (JoshuaGross) [inline haskell/src/Data/RTCM3/Ephemeris.hs]: Nice, I'm really glad you're adding the glonass ephemeris here! Meta-comment is that it'd be nice to split this into multiple structures, if possible, but it looks like there may not be a clean, sensible way of doing that.

File: haskell/src/Data/RTCM3/Ephemeris.hs [hunk 2]
```diff
@@ -202,3 +207,244 @@ instance Binary Msg1019 where
...
+  -- Units: days  Scale factor: 1
+  , _glonassEphemeris_mp4 :: Bool
+  -- ^ GLONASS-M P4 word
+  , _glonassEphemeris_mft :: Word8
+  -- ^ GLONASS-M predicted satellite user range accuracy at time @t_b@
+  , _glonassEphemeris_mnt :: Word16
+  -- ^ GLONASS_M calendar number of day within four-year interval
+  -- starting from the first of January in a leap year
+  -- Units: days  Scale factor: 1
+  , _glonassEphemeris_mM :: Word8
```
Comment (JoshuaGross) [inline haskell/src/Data/RTCM3/Ephemeris.hs]: Explicitly use a sum type here - `SatTypeGlonassM | SatTypeNonGlonassM`

File: haskell/src/Data/RTCM3/Ephemeris.hs [hunk 3]
```diff
@@ -202,3 +207,244 @@ instance Binary Msg1019 where
...
+    B.putWord8    6  _glonassEphemerisHeader_sat
+    B.putWord8    5  _glonassEphemerisHeader_channel
+
+data GlonassEphemeris = GlonassEphemeris
+  { _glonassEphemeris_almanacHealth :: Bool
+  -- ^ @C_n@ word
+  , _glonassEphemeris_healthAvailability :: Bool
+  -- ^ Is GLONASS almanac health available?
+  , _glonassEphemeris_p1 :: Word8
+  -- ^ GLONASS P1 word
```
Comment (JoshuaGross) [inline haskell/src/Data/RTCM3/Ephemeris.hs]: Does it make sense to add a little more context for what `P1` and `P2` are?

File: haskell/src/Data/RTCM3/Ephemeris.hs (hunk 1 above)
Comment (peddie) [inline haskell/src/Data/RTCM3/Ephemeris.hs]: If you have any specific suggestions, I'm open to it.  I thought about splitting out the GLONASS-M fields and putting them behind a `Maybe`, gated by `GLONASS-M M` and doing the same for the "additional data" fields, but @ljbade suggested it's better to keep things flat.

File: haskell/src/Data/RTCM3/Ephemeris.hs (hunk 3 above)
Comment (peddie) [inline haskell/src/Data/RTCM3/Ephemeris.hs]: I'll gladly add any context you can provide!  This is all the information provided in the RTCM document.

**Comment by peddie** [comment]
Must I also bump the package version if I want to be able to use the GLONASS ephemeris message in `gnss-converters`?

File: haskell/main/rtcm32json.hs [hunk 4]
```diff
@@ -1,3 +1,7 @@
+{-# OPTIONS_GHC -Wall #-}
```
Comment (mfine) [inline haskell/main/rtcm32json.hs]: `-Wall` is in the cable file already - can remove this. We don't need to put it on top of every file.

File: haskell/src/Data/RTCM3/Antennas.hs [hunk 5]
```diff
@@ -292,9 +297,7 @@ $(deriveJSON defaultOptions {fieldLabelModifier = fromMaybe "_msg1008_" . stripP
 instance Binary Msg1008 where
   get = B.runBitGet $ do
     _msg1008_descriptor    <- getBits 0
-    -- ^ Antenna descriptor.
```
Comment (mfine) [inline haskell/src/Data/RTCM3/Antennas.hs]: Can you leave these comments or improve them.

Comment (peddie) [inline haskell/src/Data/RTCM3/Antennas.hs]: These comments are just copies of the ones that appear in the data type definition.  They break haddock.  I'm not sure it's worth adding a comment about what the field is every place we use it.  Am I misunderstanding something?

File: haskell/main/rtcm32json.hs (hunk 4 above)
Comment (peddie) [inline haskell/main/rtcm32json.hs]: OK, will do.

File: haskell/src/Data/RTCM3/Antennas.hs [hunk 6]
```diff
@@ -1,3 +1,8 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE RecordWildCards #-}
```
Comment (mfine) [inline haskell/src/Data/RTCM3/Antennas.hs]: I don't think `RecordWildCards` are actually being used.

Comment (peddie) [inline haskell/src/Data/RTCM3/Antennas.hs]: They are used in the `BinaryBit` instance definitions for `AntennaReference` on lines 72 and 74.

Comment (peddie) [inline haskell/src/Data/RTCM3/Antennas.hs]: Also on lines 103 and 105 for `ExtAntennaReference`.

File: haskell/src/Data/RTCM3/Ephemeris.hs [hunk 7]
```diff
@@ -202,3 +207,244 @@ instance Binary Msg1019 where
     putBits 0 _msg1019_ephemeris
 
 $(deriveRTCM3 ''Msg1019)
+
+data GlonassEphemerisHeader = GlonassEphemerisHeader
```
Comment (mfine) [inline haskell/src/Data/RTCM3/Ephemeris.hs]: Can you rename the `EphemerisHeader` `GpsEphemerisHeader`? To avoid confusion between the two headers.

Comment (peddie) [inline haskell/src/Data/RTCM3/Ephemeris.hs]: Sure, will do.

File: haskell/src/Data/RTCM3/Antennas.hs (hunk 5 above)
Comment (mfine) [inline haskell/src/Data/RTCM3/Antennas.hs]: Oh never mind, these shouldn't have been there in the first place - it's not a data field, it's a Binary instance. Sorry about that. Definitely should be removed.

**Pull request merged**